### DISCUSSION TITLE
Seamless looping video

### DISCUSSION THREAD

**Discussion by krisxsee** [iina/iina#3673] [Ideas]
When we watched movies from DVD, it consisted of several parts of .TS files, but they played absolutely seamlessly. Is it possible to implement seamless playing in a File Loop / Playlist Loop? Pause is especially evident in video with an audio playback and I can't find a smooth solution on the Mac. For example: playing looped nft on TV in living room or Presenting for friends or co-workers a short-video. Thanks!

**Comment by low-batt** [comment]
Do you experience the same issue using the [mpv player](https://mpv.io/)?

The mpv media player is primarily a "command line" player:
> mpv is a free (as in freedom) media player for the command line

It has an "on screen controller" instead of a GUI:
> While mpv strives for minimalism and provides no real GUI, it has a small controller on top of the video for basic control. 

That project publishes a library other applications can use:
> A straightforward C API was designed from the ground up to make mpv usable as a library and facilitate easy integration into other applications.

Which brings us to IINA. IINA is one of the [GUI frontends](https://github.com/mpv-player/mpv/wiki/Applications-using-mpv#gui-frontends) for mpv.  IINA relies upon libmpv for video and audio. Therefore, for any problems related to video or audio the first question for the IINA project is whether or not the problem is reproducible using the mpv player. If it is then for IINA it is what is referred to as an "upstream" problem and needs to be entered as an issue with the mpv project: https://github.com/mpv-player/mpv/issues

In a very quick test, running the mpv player from the terminal with file looping configured:
```text
/Applications/mpv.app/Contents/MacOS/mpv --file-loop=inf ~/Movies/clip.mp4
```

The transition when playing is restarted from the beginning is rough and especially noticeable in the audio when playing a clip that is not silent around the transition. I think that is what you are reporting.

**Reply by krisxsee** [reply]
Thanks! mpv player [stolendata] works really well! 
The 2160x2160 looped video plays on 99% seamless in File Loop/Playlist Loop modes. And the 1536x1536 video plays 100% seamlessly on my MacBook 560.

**Comment by low-batt** [comment]
When playing a movie works in the mpv player and glitches in IINA there are a couple of things to check before digging into the IINA code. From the [mpv manual](https://mpv.io/manual/stable/#video):
> **--hwdec=\<api>**
> Specify the hardware video decoding API that should be used if possible. Whether hardware decoding is actually done depends on the video codec. If hardware decoding is not possible, mpv will fall back on software decoding.
> 
> Hardware decoding is not enabled by default, because it is typically an additional source of errors. It is worth using only if your CPU is too slow to decode a specific video.

That option has many potential settings. Without using advanced settings IINA only supports these:
>no:	always use software decoding (default)
auto:	forcibly enable any hw decoder found (see below)
auto-copy:	enable best hw decoder with copy-back (see below)

In IINA this is the `Hardware decoder` preference in the `Video/Audio` section. It has the following potential settings that correspond to the above mpv configuration:
- Disabled
- Auto
- Auto (Copy)

**BUT** unlike mpv which defaults to "no/Disabled", IINA defaults to "auto/Auto". So the next test to do is to disable hardware video decoding in IINA and enable it is the mpv player and see if there is any change in behavior. If IINA starts working and mpv starts showing the problem then we'd suspect the hardware drivers are to blame.

The second "big picture" issue is the version of mpv. I'm gussing you tested with the latest mpv release which is 0.34.1. The version IINA is using can be found on the `About IINA` window under the version of IINA. On Intel the current release is using 0.32.0, on M1 0.33.0. So the version of mpv being used could be the difference.

To test this you could find a corresponding old version of the mpv player and see if that shows the problem. Or you could test [IINA+](https://github.com/iina-plus/iina) which is using 0.34.1 and see if it works the same as the mpv player. IINA+ is an unofficial experimental fork of IINA.
